pkg/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,276 @@
+func NewEvent(meta EventHandler) *Event {
+	return &Event{
+		Metadata:     meta,
+		eventResults: make(chan interface{}, 1),
+		cancelled:    make(chan struct{}),
+		stats:        eventStatistics{},
+	}
+}
+
+// WasCancelled returns whether the cancelled channel for the given Event has
+// been closed or not. Cancellation occurs if the event was not processed yet
+// by an EventQueue onto which this Event was Enqueued, and the queue is closed,
+// or if the event was attempted to be scheduled onto an EventQueue which has
+// already been closed.
+func (ev *Event) WasCancelled() bool {
+	select {
+	case <-ev.cancelled:
+		return true
+	default:
+		return false
+	}
+}
+
+// Enqueue pushes the given event onto the EventQueue. If the queue has been
+// stopped, the Event will not be enqueued, and its cancel channel will be
+// closed, indicating that the Event was not ran. This function may block if
+// there is an event being processed by the queue.
+func (q *EventQueue) Enqueue(ev *Event) <-chan interface{} {
+
+	if ev == nil {
+		return nil
+	}
+
+	// Track that event has been Enqueued.
+	q.closeWaitGroup.Add(1)
+
+	select {
+	// The event should be drained from the queue (e.g., it should not be
+	// processed).
+	case <-q.drain:
+		// Closed eventResults channel signifies cancellation.
+		close(ev.cancelled)
+		close(ev.eventResults)
+		q.closeWaitGroup.Done()
+
+		return ev.eventResults
+	default:
+		// The events channel may be closed even if an event has been pushed
+		// onto the events channel, as events are consumed off of the events
+		// channel asynchronously! If the EventQueue is closed before this
+		// event is processed, then it will be cancelled.
+
+		ev.stats.waitEnqueue.Start()
+		q.events <- ev
+		ev.stats.waitEnqueue.End(true)
+		ev.stats.waitConsumeOffQueue.Start()
+		return ev.eventResults
+	}
+}
+
+func (ev *Event) printStats() {
+	log.WithFields(logrus.Fields{
+		"eventType":                    reflect.TypeOf(ev.Metadata).String(),
+		"eventHandlingDuration":        ev.stats.durationStat.Total(),
+		"eventEnqueueWaitTime":         ev.stats.waitEnqueue.Total(),
+		"eventConsumeOffQueueWaitTime": ev.stats.waitConsumeOffQueue.Total(),
+	}).Debug("EventQueue event processing statistics")
+}
+
+// Run consumes events that have been queued for this EventQueue. It
+// is presumed that the eventQueue is a buffered channel with a length of one
+// (i.e., only one event can be processed at a time). All business logic for
+// handling queued events is contained within this function. If the Metadata
+// within an Event does not implement the EventHandler interface, the function
+// will log an error, and the Event will not be processed. If the event queue is
+// closed, then all events which were queued up, but not processed, are
+// cancelled. Any event which is currently being processed will not be
+// cancelled. It is assumed that the caller handles both cases for an Event
+// (cancel, or result of event) gracefully.
```

Comment:
nit: I don't understand how the last sentence relates to the caller running `Run()` or what the caller is supposed to do about that. Does this mean that if the caller does something wrong, the eventqueue will get stuck?

Maybe it suggests that the high-level description of how to use the `EventQueue` should be documented in steps somewhere, or maybe this last sentence can just be omitted.

Looking back across the comment above `Run()` here, I don't think it's been updated since the new approach has been written for the internals of the eventqueue `Run()`/`Stop()` handling. I know it's a bit of a pain but could you review all of the comments for at least the public functions here? IMO it's better to have no comments than comments that are old/misleading/confusing.

Reply:
Sure, will review and update the docs.